PP-29: Integrate firebase auth services

richardyjtian opened the pull request:
Firebase auth is mostly integrated with the frontend following the design on https://docs.google.com/document/d/1RdtbGEHxUexCqKS7_DNWQT_68gSe7yL6uEvgYlgNCjs/edit. Services include account registration, login, sign out, password reset from email, and password change.

Account registration flow still needs to be completed in the backend in RegistrationService.tsx after a user is successfully registered in firebase (PP-36). Additional services to be added in the future include account deletion, and email change.

Resolves PP-35, PP-37, PP-41, PP-42, PP-43, PP-44, PP-45

yunishimura0716 commented on src/components/account/LoginCard.tsx:
```diff
@@ -28,24 +31,32 @@ interface Props {
+
+  const onFinish = (data: { email: string; password: string; staySignedIn: boolean }) => {
+    firebase
+      .doSignInWithEmailAndPassword(data.email, data.password, data.staySignedIn)
```
This is kind of reminder. As we discussed before, the user have to request `/users/firebase/:firebaseId` after the user succeed to sign in by firebase.

ericrkuo replied:
@yunishimura0716 I will be working on the sign in / sign up auth flow in new PR connecting it to the BE and following your auth flow doc!

yunishimura0716 commented on src/contexts/AuthUserContext/index.tsx:
```diff
@@ -0,0 +1,40 @@
+  const [authUser, setAuthUser] = useState<AuthUser | null>(null);
+  const firebase = useFirebase();
+
+  useEffect(() => {
```
I just want to ask you about `onAuthStateChanged`. Is this called every time the user `signIn` or `signUp` ?

richardyjtian replied:
onAuthStateChanged is called every time the user signs in, but the user will automatically be signed in after signing up for an account too

yunishimura0716 replied:
Sounds good!

yunishimura0716 reviewed (commented):
Great work, Richard! Thank you so much! The implementation makes sense to me. Some codes for auth flow with Backend is not implemented yet. If you have question for that, let me know.

ericrkuo commented on src/services/RegistrationService.ts:
```diff
@@ -0,0 +1,19 @@
+      .doCreateUserWithEmailAndPassword(data.email, data.password)
+      .then(() => {
+        // TODO: Register user in backend
+        // https://promopal.atlassian.net/browse/PP-36
```
I think PP-36 has already been implemented, could we connect this to the BE?

andrewzulaybar commented on src/services/FirebaseService.ts:
```diff
@@ -0,0 +1,75 @@
+};
+
+class Firebase {
+  auth: firebase.auth.Auth;
```
Can we make this private?

andrewzulaybar commented on src/services/FirebaseService.ts:
```diff
@@ -0,0 +1,75 @@
+
+  // *** Auth API ***
+  // Creates a user and signs the user in
+  doCreateUserWithEmailAndPassword = (email: string, password: string) => {
```
Minor thing I forgot to mention last time, let's use regular function declarations instead of arrow functions for functions inside classes so:
```suggestion
  doCreateUserWithEmailAndPassword(email: string, password: string) {
```

andrewzulaybar replied:
Also, just merged in a change to the linter that forces us to add types to return values of functions so you'll probably have to add that!

andrewzulaybar reviewed (approved):
This is really great, excited to get this in! I'll work on handling the error propagation this week. Just added a few comments.

Last few things we'll probably have to do is restrict access for certain pages if a user is not logged in, which we can work out this week. If there are other things you can think of that needs to be done that aren't in tickets already, let me know and we can create tickets for them + distribute among us this week

richardyjtian commented:
I made some changes with the merge commit including adding the ability for users to change their email registered on firebase. @andrewzulaybar, I modified some of your code in UserService.ts - could you re-review this file when you have time? I also moved RegistrationService.ts into UserService.ts @ericrkuo for when you connect registration to the backend.

ericrkuo commented on src/services/FirebaseService.ts:
```diff
@@ -0,0 +1,100 @@
+    password: string,
+    staySignedIn: boolean
+  ): Promise<firebase.auth.UserCredential> {
+    if (staySignedIn) {
+      this.auth.setPersistence(firebase.auth.Auth.Persistence.LOCAL);
+    } else {
+      this.auth.setPersistence(firebase.auth.Auth.Persistence.SESSION);
+    }
```
Do you mind explaining how this will work?

richardyjtian replied:
So session means the user will be logged out after the window is closed, and local means the user will stay signed in when the window is closed and re-opened.

ericrkuo commented on src/services/FirebaseService.ts:
```diff
@@ -0,0 +1,100 @@
+  doReauthenticateUser(password: string): Promise<firebase.auth.UserCredential> {
+    const authUser = this.auth.currentUser;
+    // TODO: Redirect to 404 page if a user or user email does not exist
+    if (authUser === null) {
+      return Promise.reject(new Error('No user is logged in.'));
+    }
+    if (authUser.email === null) {
+      return Promise.reject(new Error('User email does not exist.'));
+    }
```
Just to get my understanding better, in what situations will `authUser` and `authUser.email` be null?

richardyjtian replied:
This should not ever happen in this function because it is called in the account details page where the user will be logged in and requesting to change their password or email. This is here more or less to address typescript errors and unauthorized users somehow access the account page.

ericrkuo commented on src/contexts/AuthUserContext/index.tsx:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Use this function to access the context object.
+ */
+export function useAuthUser(): AuthUser | null {
+  const context = useContext(AuthUserContext);
+  return context;
+}
```
```suggestion
export function useAuthUser(): AuthUser | null {
  return useContext(AuthUserContext);
}
```

ericrkuo commented on src/components/navigation/NavigationBar.tsx:
```diff
@@ -29,7 +31,10 @@ export default function NavigationBar(): ReactElement {
+  const firebase = useFirebase();
+  const authUser = useAuthUser();
 
+  // TODO: isActive is not highlighting the active page when the browser back button is pressed
   const isActive = (key: Pages): CSSProperties => ({
     fontWeight: current === key ? 'bold' : 'normal',
```
Nice catch, could we create a bug ticket and link it?

richardyjtian replied:
Yup, I've created https://promopal.atlassian.net/browse/PP-101 for it

richardyjtian commented on src/services/UserService.ts:
```diff
@@ -93,21 +94,51 @@ class UserService {
+   * @param data - Data to update the user with
+   */
+  public async updateUser(firebase: Firebase, data: UserInputData): Promise<void> {
+    return firebase.doEmailUpdate(data.password, data.email).then(() => {
+      // Update user in BE
+      const url = Routes.USERS.UPDATE(this.userId);
+      const user: Partial<User> = {
+        email: data.email,
+        firstName: data.firstName,
+        lastName: data.lastName,
+        username: data.username,
+      };
+      return axios
+        .patch(url, user)
+        .then(({ data }: AxiosResponse<UpdateUserResponse>) => {
+          if (isError<UpdateUserResponse>(data)) {
+            return Promise.reject(data);
+          }
+          return Promise.resolve();
+        })
+        .catch((err: Error) => Promise.reject(err));
+    });
```
Yes only tracking the email in FB sounds good. We can also approach it like in the user registration design, where if it succeeds in FB, we make a request to BE, and if it fails in BE, we abandon in FB.

ericrkuo replied:
No longer needs this because we are not storing email in our BE!

ericrkuo commented:
@richardyjtian sorry I must have broken the build, didn't realize we need to include it in the dependency array https://github.com/ubclaunchpad/PromoPal-frontend/runs/2265756620

richardyjtian commented:
> @richardyjtian sorry I must have broken the build, didn't realize we need to include it in the dependency array https://github.com/ubclaunchpad/PromoPal-frontend/runs/2265756620

No problem, I'll fix it before I merge

ericrkuo commented:
> > @richardyjtian sorry I must have broken the build, didn't realize we need to include it in the dependency array https://github.com/ubclaunchpad/PromoPal-frontend/runs/2265756620
> 
> No problem, I'll fix it before I merge

Ah okay, thanks Richard! Would you mind tagging Andrew and I for re-review once more when you're ready since you'll be making some crucial changes?

andrewzulaybar commented on src/services/PromotionService.ts:
```diff
@@ -118,17 +119,18 @@ export async function queryPromotions(filters: FilterOptions, sort?: Sort): Prom
-    }
-    promotionQueryDTO.push(queryParams);
-  }
+  // todo: commented out until BE PR for sort is in
```
We can uncomment this now since https://github.com/ubclaunchpad/PromoPal-backend/pull/127 should be ready to go!

richardyjtian replied:
I left this commented out actually, as there seemed to be an error in push() and I didn't want to mess things up haha

ericrkuo replied:
Good point, we should merge the BE PR first before uncommenting. 

The reason why we don't need push is that I changed it from type `Record<string, string>` to type `PromotionQueryDTO`, so now we should be able to do
```
promotionQueryDTO.sort = sort
promotionQueryDTO.lat = lon
promotionQueryDTO.lon = lon
```

ericrkuo commented on src/components/upload-promotion/UploadPromotionForm.tsx:
```diff
@@ -46,10 +46,16 @@ export default function UploadPromotionForm(): ReactElement {
   const [datesEffective, setDatesEffective] = useState<PromotionDates>({ start: '', end: '' });
   const [times, setTimes] = useState<PromotionTimes>({});
 
+  // TODO: https://promopal.atlassian.net/browse/PP-80
+  if (!authUser) {
+    return <p>Error: No user is logged in.</p>;
+  }
```
We're guaranteed that auth User is defined right, since the only way to get into this route is for auth user to be defined, otherwise it takes us to login form right

richardyjtian replied:
Right, these if (!authUser) checks are only to see if a user illegally accesses the page
